[incubator/elasticsearch] Using DNS discovery instead of kubernetes plugin

mikn opened the pull request:
This PR removes the novel kubernetes discovery plugin and leverages DNS unicast discovery through a service that exposes the masters transport port instead.
This allows you to run any vanilla elasticsearch image. I have picked blacktop/elasticsearch as it is a slim Alpine-based image that releases new versions within a day or two, but any user of these charts may pick any image they deem suitable.
This also upgrades to ES6, where the fsGroup security setting is required as otherwise Elasticsearch won't boot.

mikn committed:
Using DNS discovery instead of kubernetes plugin

mikn committed:
Updating README to warn users about update to ES6

timstoop commented:
Very nice solution for discovery, no plugin needed! I checked the code and it looks fine. The warning in the README is an important one, too bad that we cannot make helm give a warning when upgrading to a certain version from an older version. Then again, this is incubator, so people should probably expect things to do this sometimes? Are there guidelines regarding that?

timstoop commented on incubator/elasticsearch/values.yaml:
```diff
@@ -4,12 +4,13 @@
 image:
-  repository: "centerforopenscience/elasticsearch"
-  tag: "5.4"
+  repository: "blacktop/elasticsearch"
```
Any particular reason why you wouldn't use the official upstream images? I'd say a generic helm chart should prefer the official images, probably.

rendhalver replied:
I was wondering the same thing.

mikn replied:
Main reason is that the official images install a trial version of X-Pack and are about 3x the size of this image. I was however going back and forth between the official image and this one. This pick was made as a personal recommendation rather than "I want to make sure I'm on good terms with upstream companies". I'm willing to leave the pick of default image up to a democratic process. :)

rendhalver replied:
Sounds like a good reason to me.

audip replied:
I'd prefer to use official docker images: https://www.docker.elastic.co/

timstoop reviewed (approved):
The changes look good to me. Only thing I'm not sure about is whether we want to use official images or not, as no plugin is necessary anymore.

audip commented:
Should `client` files be renamed to `coordinator`, as upstream calls it?
https://discuss.elastic.co/t/client-node-and-coordinating-nodes/82384

audip commented:
Also, the `client-svc.yaml` shouldn't have to expose transport port, only masters should have 9300 exposed:
```
    - name: transport
      port: 9300
      targetPort: transport
```

mikn commented:
I agree about the client service port, and I did initially yank it in my PR, but I decided to roll that back as it wasn't directly related to the proposed change here.
I'll update with removing the SA, yanking the transport port from the client service and changing to the official image.
I've also gone back and forth on whether the client nodes should be renamed, though. As mentioned in the reply to the post you link to, they're only called coordinating nodes when they do not have any of the features enabled (not master, data nor ingest), so they're just participating in the cluster to be able to vote in master elections (think arbiter nodes in MongoDB). The "client nodes" in this chart, however have ingest enabled. But they're also not only for ingestion, as they're used for both read/write into the cluster.
I chose not to rename them because of this ever so slight ambiguity. :)

mikn committed:
Updating according to PR comments

- Remove SA and RBAC role bindings (document that it isn't needed)
- Remove transport port from client service
- Change to official image

audip reviewed (commented):
Thanks for the update and it is much better than before. I haven't tested this PR in a test cluster, hence not marking it as approved, but looks good to me.

mikn committed:
Changing to OSS image to not include X-Pack trial per default

audip commented on incubator/elasticsearch/values.yaml:
```diff
@@ -4,12 +4,13 @@
 image:
-  repository: "centerforopenscience/elasticsearch"
-  tag: "5.4"
+  repository: "docker.elastic.co/elasticsearch/elasticsearch-oss"
```
This is a cool image. 👍

icereval replied:
The `pre-stop-hook.sh` currently relies on the json query tool [`jq`](https://github.com/kubernetes/charts/blob/master/incubator/elasticsearch/templates/configmap.yaml#L122) to parse the k8s api responses. I'm on board with sticking with the default image, and historically had only kept the very [minimally wrapped](https://github.com/CenterForOpenScience/docker-library/blob/master/elasticsearch/5.4/Dockerfile#L5) official image due to the need for the [fabric8 plugin](https://github.com/CenterForOpenScience/docker-library/blob/master/elasticsearch/5.4/Dockerfile#L18). The script should be updated to use an inline `python` command to perform the same operations.

icereval replied:
A thought, but we'll likely need to wrap the image again when we get around to [adding TLS support](https://github.com/CenterForOpenScience/docker-library/blob/master/elasticsearch/5.5/tls/Dockerfile#L19), it is not included in X-Pack :-(.

rendhalver commented:
Hi gang.
We were unable to boot this up with the official elasticsearch-oss image.
We had to turn off the xpath options.
Do you want a PR on this PR that will fix that?
We basically added an extra section to appVersion 6.x without the path options.

This is the commit I added to fix it.
https://github.com/powerhome/kubernetes-charts/commit/2a15952ce23f3f957f93f664b462e66393e00a57

timstoop commented:
Maybe @mikn can pull your commit into his PR branch? I haven't tested it myself, but I can imagine the options simply do not exist when X-Pack is not installed?

timstoop commented:
Actually, I guess the X-Pack options should just be configurable? As in, whether you want them at all or not?

rendhalver commented:
It looks like they aren't there when x-pack isn't installed.
I am not sure why ES doesn't just ignore them if the plugin isn't installed.

mikn commented:
Hah, sorry about that people. I'll push a fix ASAP. That's what I get for not testing when changing image!

mikn committed:
Disable X-Pack options with a flag and update image to 6.0.1

rendhalver commented on incubator/elasticsearch/templates/configmap.yaml:
```diff
@@ -38,11 +38,13 @@ data:
         ping.unicast.hosts: ${DISCOVERY_SERVICE:}
         minimum_master_nodes: ${MINIMUM_MASTER_NODES:2}
 
+{{- if .Values.cluster.xpackEnable }}
```
That's a better plan than my idea. :) 👍

rendhalver commented on incubator/elasticsearch/values.yaml:
```diff
@@ -5,12 +5,15 @@ appVersion: "5.4"
 image:
   repository: "docker.elastic.co/elasticsearch/elasticsearch-oss"
-  tag: "6.0.0"
+  tag: "6.0.1"
```
Another thing I noticed with our deploy was that Kibana 5.4.3 complained that not all parts of our Elk stack were at the same version. I am not sure that's still the case with the 6.x series.

rendhalver commented:
We are using this branch in our local stack so it's getting tested there.
We will let you know if we spot anything else. :)

rendhalver commented:
What do we need to get this fixed and merged?

timstoop commented:
A rebase at the very least :) But I think there's interest in Slack's #elk-charts to have this merged then, from what I can see.

monotek commented:
I'd love to see this merged, to be able to use ES 6.1.1 :-)

mikn committed:
Updating image to 6.1.1

mikn commented:
/assign @lachie83

mattfarina commented:
/cc @simonswine @icereval

pavelnikolov commented on incubator/elasticsearch/templates/master-statefulset.yaml:
```diff
@@ -71,20 +72,14 @@ spec:
           value: "false"
 {{- end }}
+        - name: DISCOVERY_SERVICE
+          value: {{ template "elasticsearch.fullname" . }}-discovery.{{ .Release.Namespace }}.svc.{{ .Values.cluster.kubernetesDomain }}
```
Since `{{ template "elasticsearch.fullname" . }}-discovery.{{ .Release.Namespace }}.svc.{{ .Values.cluster.kubernetesDomain }}` is repeated 3 times, I guess it's worth putting it in its own helper template. WDYT?

icereval commented on incubator/elasticsearch/README.md:
```diff
@@ -1,7 +1,13 @@
-This chart is based on the [centerforopenscience/elasticsearch](https://hub.docker.com/r/centerforopenscience/elasticsearch/) image which comes with Fabric8's great [kubernetes discovery plugin](https://github.com/fabric8io/elasticsearch-cloud-kubernetes) for Elasticsearch.
- 
+This chart uses a standard Docker image of Elasticsearch (hub.docker.com/r/blacktop/elasticsearch) and uses a service pointing to the master's transport port for service discovery.
+Elasticsearch does not communicate with the Kubernetes API, hence no need for RBAC permissions.
```
RBAC is still needed for the `pre-stop-hook.sh`. This script scales down the statefulset data pods properly. In order to avoid data loss it essentially instructs ES to isolate the host being removed, and evacuates all documents and shard replicas. If you could determine the same information w/o having to make k8s api calls, rather query DNS that would certainly be ideal.

mikn replied:
Ah, I've been looking in to the script and it looks at the desired count of the statefulset to determine whether a scale down is required (very clever!). I'll dig around a bit, but it seems highly unlikely this is information provided elsewhere...

icereval commented on incubator/elasticsearch/templates/client-deployment.yaml:
```diff
@@ -21,7 +21,8 @@ spec:
         {{- end }}
     spec:
-      serviceAccountName: {{ template "elasticsearch.fullname" . }}
+      securityContext:
```
The ES runtime should step down from `root` and execute as the `elasticsaerch` user (PID 1000), is this fixing a specific issue you were running into?

mikn replied:
As of ES 6.0 Elasticsearch requires the config directory and data directory to be owned by the ES user, this ensures the mounted volumes are owned by ES. :)

kalabiyau replied:
@mikn In some implementations of k8s (namely SUSE CaaSP) you need to explicitly grant priviledged mode to the service-account via RBAC, otherwise master won't start.

icereval commented on incubator/elasticsearch/values.yaml:
```diff
@@ -84,11 +88,6 @@ data:
   #   example: data-foo
-
-## Install Default RBAC roles and bindings
-rbac:
```
same as above, `pre-stop-hook` script is still relying on k8s api calls.

icereval reviewed (changes requested):
I've tested a simple deployment with GKE, and the DNS changes appear to work great!

As a suggestion, it might be worth moving the image, rbac and associated script changes mentioned to another PR, as they will cause this to incrase in scope.

I've updated the COS repo with a wrapped ES 6.1.1 image, though I left X-Pack in the base image since it is toggled off by default.

Thanks for the PR!

mikn commented:
Thanks for the review! I'll make sure to address the issues asap~
As for TLS support, yes. Most likely will require a wrapped image, I did implement it myself using init containers initially for plugin install, however. :)

icereval commented:
@mikn, what cloud provider have you tested with?

mikn commented:
I've only tested on a kops installation on AWS and Minikube.

rendhalver commented:
We have been running this locally on our Tectonic cluster for a while.

mattfarina commented:
@icereval note, since you're a maintainer I'm holding off reviewing until you give it a LGTM.

@mikn it appears this may need to be rebased from master.

icereval commented:
@mattfarina, sounds good, we are still working through some RBAC + script changes included and/or potentially splitting those changes out of the core DNS discovery change here.

@mikn & @rendhalver, great to hear its working well on AWS, just wanted to double check as I've been using GKE.

mikn commented:
@icereval I've done some initial testing and it seems I can detect scale-down events by doing `host -t SRV [HEADLESS SERVICE]`. I haven't confirmed I can detect in the scale-down script yet, and this would require adding in a headless service for the data nodes for this sole purpose.
I'm honestly torn whether it'd be worth it or not, so what do you guys think?
I'll have this PR ready for next review before next week, for the eager watchers. :)

icereval commented:
@mikn my preference would be to leverage RBAC and the k8s api. Eventually this could be replaced by an operator pattern following the k8s api's. e.g. https://github.com/upmc-enterprises/elasticsearch-operator

rendhalver commented:
@icereval We run a local Kubernetes cluster on bare metal if that makes any difference.

icereval commented:
@rendhalver, thanks for posting, and thats good to know. My main concern was we didn't switch to a DNS based unicast networking approach and find out others would/might immediately find otherwise this approach might not work as well as we thought. However I'd say with GKE, AWS and especially an on-site approach we've done a good bout of validation and testing. Thanks all!

icereval commented:
@mikn e.g. of narrowed down changes in a fork @ https://github.com/CenterForOpenScience/helm-charts/commit/319d7b7bf71325c4ab7bcf27212c0b902dba7b04

KongZ commented:
@mikn @icereval  I created a new PR to remove JQ dependency from `pre-stop-hook.sh`

https://github.com/kubernetes/charts/pull/3568

monotek commented:
Whats blocking this?

unguiculus committed:
Merge branch 'master' into es_svc_discovery

unguiculus commented:
/ok-to-test

unguiculus commented:
So, can this be merged now? 
@simonswine @icereval

unguiculus commented:
/assign

monotek commented:
@mikn 
Can you resolve the merge conflicts please?

unguiculus committed:
Merge branch 'master' into es_svc_discovery

Xuhao commented:
Any update for this PR?

unguiculus commented:
Any chance, this gets fixed?

```
I0413 20:42:50.779] [2018-04-13T20:24:34,854][WARN ][o.e.d.z.UnicastZenPing   ] [elastic-8523-1-elasticsearch-client-74c4c5d9cb-kbbq7] failed to resolve host [elastic-8523-1-elasticsearch-discovery.presubmit-2889-8523.svc.cluster.local]
I0413 20:42:50.779] java.net.UnknownHostException: elastic-8523-1-elasticsearch-discovery.presubmit-2889-8523.svc.cluster.local: Name or service not known
I0413 20:42:50.780] 	at java.net.Inet4AddressImpl.lookupAllHostAddr(Native Method) ~[?:1.8.0_151]
I0413 20:42:50.780] 	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928) ~[?:1.8.0_151]
I0413 20:42:50.780] 	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323) ~[?:1.8.0_151]
I0413 20:42:50.780] 	at java.net.InetAddress.getAllByName0(InetAddress.java:1276) ~[?:1.8.0_151]
I0413 20:42:50.780] 	at java.net.InetAddress.getAllByName(InetAddress.java:1192) ~[?:1.8.0_151]
I0413 20:42:50.781] 	at java.net.InetAddress.getAllByName(InetAddress.java:1126) ~[?:1.8.0_151]
I0413 20:42:50.781] 	at org.elasticsearch.transport.TcpTransport.parse(TcpTransport.java:917) ~[elasticsearch-6.1.1.jar:6.1.1]
I0413 20:42:50.781] 	at org.elasticsearch.transport.TcpTransport.addressesFromString(TcpTransport.java:872) ~[elasticsearch-6.1.1.jar:6.1.1]
I0413 20:42:50.781] 	at org.elasticsearch.transport.TransportService.addressesFromString(TransportService.java:699) ~[elasticsearch-6.1.1.jar:6.1.1]
I0413 20:42:50.782] 	at org.elasticsearch.discovery.zen.UnicastZenPing.lambda$null$0(UnicastZenPing.java:213) ~[elasticsearch-6.1.1.jar:6.1.1]
I0413 20:42:50.782] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_151]
I0413 20:42:50.782] 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:568) [elasticsearch-6.1.1.jar:6.1.1]
I0413 20:42:50.782] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_151]
I0413 20:42:50.782] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_151]
I0413 20:42:50.783] 	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_151]
I0413 20:42:50.783] [2018-04-13T20:24:37,863][WARN ][o.e.d.z.ZenDiscovery     ] [elastic-8523-1-elasticsearch-client-74c4c5d9cb-kbbq7] not enough master nodes discovered during pinging (found [[]], but needed [2]), pinging again
```

rendhalver commented:
@mikn Need a hand with getting this sorted?
I may have some spare time on the weekend I can devote to charting.

knorby commented on incubator/elasticsearch/templates/configmap.yaml:
```diff
@@ -10,7 +10,7 @@ metadata:
+
     node.data: ${NODE_DATA:true}
     node.master: ${NODE_MASTER:true}
 {{- if hasPrefix "5." .Values.appVersion }}
```
I believe this is the only real issue right now. App version is now 6.x. Fixing this line and the same thing below should gets tests to pass, I believe.

rendhalver replied:
I am fixing this in #5663

rendhalver commented:
How would I push changes into this branch so we can get this fixed?

monotek commented:
i guess pull request against: https://github.com/mikn/charts/tree/es_svc_discovery

timstoop commented:
Unless @mikn is no longer interested in doing the rest of the work (which is fine, people move on all the time), then it might be worthwhile to clone this PR and resubmit it from your own branch instead. Do you still have the time/interest in working on this, @mikn?

rendhalver commented:
I am happy to redo it if you don't have the time @mikn.

rendhalver commented:
Or I guess I will just open up a new one in the meantime and we can go with that if we don't hear back in a few days.
I will make a start on that when I get a chance.

rendhalver commented:
I have managed to get the tests passing in my redo PR.
Were there any other concerns that needed addressing?

Vince-Cercury commented:
Any update?

rendhalver commented:
@VinceMD You can follow along on the redo over here #5663

davidkarlsen commented:
@mikn Should this issue be closed since https://github.com/kubernetes/charts/pull/5663 was merged?

monotek commented:
I think so...

unguiculus commented:
/close
